python/ray/serve/server.py

```diff
@@ -127,63 +111,61 @@ def _check_slo_ms(self, request_slo_ms):
+            ).format(current_path)
+            await error_sender(error_message, 404)
+            return
+
+        endpoint_name, methods_allowed = self.route_table_cache[current_path]
+
+        if scope["method"] not in methods_allowed:
+            error_message = ("Methods {} not allowed. "
+                             "Avaiable HTTP methods are {}.").format(
+                                 scope["method"], methods_allowed)
+            await error_sender(error_message, 405)
             return
 
-        endpoint_name = self.route_table_cache[current_path]
         http_body_bytes = await self.receive_http_body(scope, receive, send)
 
         # get slo_ms before enqueuing the query
-        query_string = scope["query_string"].decode("ascii")
-        query_kwargs = parse_qs(query_string)
-        relative_slo_ms = query_kwargs.pop("relative_slo_ms", None)
-        absolute_slo_ms = query_kwargs.pop("absolute_slo_ms", None)
         try:
-            relative_slo_ms = self._check_slo_ms(relative_slo_ms)
-            absolute_slo_ms = self._check_slo_ms(absolute_slo_ms)
-            if relative_slo_ms is not None and absolute_slo_ms is not None:
-                raise ValueError("Both relative and absolute slo's"
-                                 "cannot be specified.")
+            relative_slo_ms, absolute_slo_ms = self._parse_latency_slo(scope)
         except ValueError as e:
-            await JSONResponse({"error": str(e)})(scope, receive, send)
+            await error_sender(str(e), 400)
             return
 
         # create objects necessary for enqueue
         # enclosing http_body_bytes to list due to
         # https://github.com/ray-project/ray/issues/6944
         # TODO(alind):  remove list enclosing after issue is fixed
         args = (scope, [http_body_bytes])
+        headers = {k.decode(): v.decode() for k, v in scope["headers"]}
         request_in_object = RequestMetadata(
             endpoint_name,
             TaskContext.Web,
             relative_slo_ms=relative_slo_ms,
-            absolute_slo_ms=absolute_slo_ms)
-
-        actual_result = await (self.serve_global_state.init_or_get_router()
-                               .enqueue_request.remote(request_in_object,
-                                                       *args))
-        result = actual_result
-
-        if isinstance(result, ray.exceptions.RayTaskError):
-            await JSONResponse({
-                "error": "internal error, please use python API to debug"
-            })(scope, receive, send)
-        else:
-            await JSONResponse({"result": result})(scope, receive, send)
+            absolute_slo_ms=absolute_slo_ms,
+            call_method=headers.get("X-SERVE-CALL-METHOD".lower(), "__call__"))
+
+        try:
+            result = await (self.serve_global_state.init_or_get_router()
+                            .enqueue_request.remote(request_in_object, *args))
+            await Response(result)(scope, receive, send)
```

Comment:
This looks a bit awkward to me. Isn't something like this more readable? 
```suggestion
            await Response(result).send(scope, receive, send)
```

It is your choice though.